src/v/cloud_storage/materialized_resources.cc

```diff
@@ -115,84 +108,169 @@ ss::future<> materialized_resources::start() {
+    // ss::input_stream doesn't consume that much memory all the time, only
+    // if it's actively used.
+    return static_cast<ssize_t>(
+      config::shard_local_cfg().storage_read_buffer_size()
+      * config::shard_local_cfg().storage_read_readahead_count());
+}
+
+inline ssize_t projected_remote_partition_reader_memory_usage() {
+    // This value is just an estimate of a real thing. The reader contains
+    // a bunch of fileds and offset translation state which is variable size.
+    // We can't really know in advance how big it is (usually it's not big).
+    static size_t sz = remote_partition::reader_mem_use_estimate();
+    return (ssize_t)sz;
+}
+
+inline ssize_t projected_remote_segment_memory_usage() {
+    // This is an estimate. When the reader is created it's size should be
+    // checked and if it's lareger units should be acquired. If it's smaller
+    // some units should be released.
+    static constexpr size_t segment_index_projected_size = 0x1000;
+    return sizeof(remote_segment) + segment_index_projected_size;
+}
+
 materialized_manifest_cache&
 materialized_resources::get_materialized_manifest_cache() {
     return *_manifest_cache;
 }
 
-size_t materialized_resources::max_segment_readers() const {
-    return static_cast<size_t>(_max_segment_readers_per_shard().value_or(
-      _max_partitions_per_shard() * default_reader_factor));
-}
-
-size_t materialized_resources::max_partition_readers() const {
-    return static_cast<size_t>(_max_partition_readers_per_shard().value_or(
-      _max_partitions_per_shard() * default_reader_factor));
+size_t materialized_resources::max_memory_utilization() const {
+    auto max_readers
+      = config::shard_local_cfg().cloud_storage_max_segment_readers_per_shard();
+    if (max_readers.has_value()) {
+        auto max_mem = projected_remote_segment_reader_memory_usage()
+                       * max_readers.value();
+        return std::min(
+          memory_groups::tiered_storage_max_memory(), (size_t)max_mem);
+    }
+    return memory_groups::tiered_storage_max_memory();
 }
 
-size_t materialized_resources::max_segments() const {
-    return static_cast<size_t>(_max_segments_per_shard().value_or(
-      _max_partitions_per_shard() * default_segment_factor));
+size_t materialized_resources::max_parallel_hydrations() const {
+    auto max_connections
+      = config::shard_local_cfg().cloud_storage_max_connections();
+    auto n_readers = config::shard_local_cfg()
+                       .cloud_storage_max_partition_readers_per_shard();
+    if (n_readers) {
+        return std::min((unsigned)max_connections, n_readers.value());
+    }
+    return max_connections / 2;
 }
 
 size_t materialized_resources::current_segment_readers() const {
-    return _segment_reader_units.outstanding();
+    size_t n = 0;
+    for (const auto& s : _materialized) {
+        n += s.readers.size();
+    }
+    return n;
 }
 
 size_t materialized_resources::current_partition_readers() const {
```

Comment:
Is it worth renaming this to something `current_ongoing_hydrations`?

Reply:
done